### PR TITLE
Fixed `.completed` being sent after `.error` & unified notifications

### PR DESCRIPTION
**Pull request by vhbit**
In a couple of places `.completed` was sent after `.error` which breaks
contract of `Observable`. Along the way, unified how observers are being
notified: now everything goes through `on(event)` instead of
`onNext/onCompleted/onError`.

### CODE REVIEW

**Comment by vhbit** [comment]
The issue addressed here has nothing specific to `RxAlamofire`: by a contract `Observable` can generate 0 or more `.next` events, but it always ends either with `.error` or `.completed` - both of those are considered stream "terminators" and there should be just one in the end.

What I'm addressing here is that in some cases wrappers send `.error` and `.completed` afterwards, which breaks that contract. 

Now back to your questions:
1. I'm not sure that HTTP streams are supported at all ATM. But anyway, as you can see, the only places changed are there both `.error` and `.completed` were sent from: `(.next | .error), .completed` to `(.next, .completed) | .error`

2.
> The "completed" event should be ignored by RxSwift (is it not?)

I'm not sure it is ignored on RxSwift level as for example it might give unexpected results when using `materialize`. So I think it's better to keep a contract and send a correct sequence of events instead of relying on some kind of magic happening behind the scenes.

@kzaher please give an advice if in general we should care about it or that's too pedantic :smile:

3. The only reason is to have unified code as it allows to do things a bit faster. For example, here I used search and replaced all placed with `on(event)`. While replacing I've found that `onCompleted/onError` is used in some places and had to go for another round of search/replace. So I've switched to `RxSwift` approach, i.e. `on(event)`.

**Comment by freak4pc** [comment]
I personally prefer the shorthand onNext, onError, onComplete but I guess i'm not too "religious" on that so all good 👍 

About the completed event I believe that is the contract of *RxSwift itself*, and not necessarily any specific library. You can't always know when a stream of events would end and not every next event should necessarily end with a completed (you might have some people depending on multiple next events from a streamed/buffered request or a different edge case you're not using on a daily basis). 

As you can see Next/Error events are not emitted by the Observable implementation and that is the actual contract of RxSwift itself: 

9b3ac1d3-1211-4214-8596-bf3eaa992a90

**Comment by vhbit** [comment]
@freak4pc 

> You can't always know when a stream of events would end and not every next event should necessarily end with a completed 

Again, it's not about sending `.completed` every time, but about sending either `.completed` or `.error`

> As you can see Next/Error events are not emitted by the Observable implementation and that is the actual contract of RxSwift itself:

You're comparing specifics of `PublishSubject` implementation with RxAlamofire, which is incorrect, such behavior is *not* present in Alamofire, in fact this PR actually brings it to be on par with `PublishSubject` and other good RxSwift citizens.

 Without this patch the code similar to following:

```swift
_ = request(.get, stringURL)
    .flatMap { request in
        return request.validate(statusCode: 200..<300)
        .validate(contentType: ["text/json"])
            .rx.json()
    }
    .observeOn(MainScheduler.instance)
    .subscribe(onCompleted: { print("completed \($0)") }, 
               onError: { print("error \($0)") })
```

might give unexpected result "completed ... error ..."

As you can see, now it all depends on what exactly client side does in this case and while it's very unlikely, it might happen that completed or error blocks will do exclusive side-effects. Maybe some clients will show a success HUD and error HUD, maybe some clients will process or delete file and so on, nobody knows, but we can prevent it.

**Comment by freak4pc** [comment]
I’m at WWDC this week so probably won’t have time to look but thanks for the write up! I’ll try looking over the weekend @vhbit

File: Sources/RxAlamofire.swift [hunk 1]
```diff
@@ -775,11 +775,11 @@ extension Reactive where Base: DataRequest {
                 case .success(let result):
                     if let httpResponse = packedResponse.response {
                         observer.on(.next(httpResponse, result))
+                        observer.on(.completed)
                     }
                     else {
                         observer.on(.error(RxAlamofireUnknownError))
                     }
-                    observer.on(.completed)
```
Comment (sergdort) [inline Sources/RxAlamofire.swift]: Good catch 👍

Comment (freak4pc) [inline Sources/RxAlamofire.swift]: Is it not an issue if the stream needs to send more than 1 value ? like mentioned above?

Comment (sergdort) [inline Sources/RxAlamofire.swift]: Yeah, I was thinking about it as well. The problem here is that I'm not sure if there is a way to distinguish if this is streaming connection. 
For example I ones used it for [Firebase](https://firebase.google.com/docs/reference/rest/database/#section-streaming) (because they do not support tvOS) and the only difference is in `Accept` header field

Comment (freak4pc) [inline Sources/RxAlamofire.swift]: Actually the [Alamofire 4.0 Migration Guide](https://github.com/Alamofire/Alamofire/blob/master/Documentation/Alamofire%204.0%20Migration%20Guide.md#benefits-of-upgrading) claims 

> New Request Types: include DataRequest, DownloadRequest, UploadRequest and **StreamRequest** that implement specialized progress, validation and serialization APIs and behaviors per Request type.

Would that mean DataRequest doesn't work with streamed request here? If yes this fix should be valid.

Comment (vhbit) [inline Sources/RxAlamofire.swift]: This fix is orthogonal to streaming support as:
- if it was supported - it continues to support it
- if it wasn't supported - it continues to not support it

but I really think that streaming was never supported in RxAlamofire

Comment (freak4pc) [inline Sources/RxAlamofire.swift]: I don't think it's orthogonal :) 
If it was supported (multiple next events), this will break it. 

But - to be honest I think it was never supported with this method and actually this would drive us closer to moving to Single so I'm all for supporting the most common use case. 

👍 from me

Comment (vhbit) [inline Sources/RxAlamofire.swift]: @freak4pc 

> If it was supported (multiple next events), this will break it

well, all the changes are here, I'd be happy to see how it breaks it with reference to the lines and explanation how it does it

Comment (freak4pc) [inline Sources/RxAlamofire.swift]: @vhbit line 778 emits completion immediately after next. Meaning if it was a streamed response you only get a single emission regardless and dispose after it. **if** streamed responses were supported there that’s how it could possibly break it.

But as I said I think we should probably not worry about it.

Comment (vhbit) [inline Sources/RxAlamofire.swift]: @freak4pc 

> line 778 emits completion immediately after next

that's true but if you check code before patch, you'll see that it got to line 778 from line 782, which was executed immediately after next before, i.e. my statement still holds - it doesn't change anything regarding streaming support

**Comment by freak4pc** [comment]
Ok gonna leave one last comment here and wait for any "Against" notes
@sergdort any reason to not merge it? If there's no opposition I'm good to merge.

**Comment by sergdort** [comment]
Yeah, good improvement 👍
